cohortextractor/sqlalchemy_types.py

```diff
@@ -1,12 +1,52 @@
 import sqlalchemy.types
+from sqlalchemy.types import Boolean, Float, Integer, String, Text
+
+
+__all__ = [
+    "Boolean",
+    "Date",
+    "DateTime",
+    "Float",
+    "Integer",
+    "String",
+    "Text",
+    "TYPES_BY_NAME",
+]
+
+# SQLAlchemy has different coercion rules for custom types vs built-in types. For
+# instance, if you compare a date and a string (using the standard types for both)
+# SQLAlchemy will decide how the type coercion should be done without ever checking
+# whether the current dialect supplies its own date type with its own coercion rules.
+# For our use case this is undesirable as we want to be able to control type coercion
+# behaviour. In particular we want to make sure that if a date is every compared to a
+# string in MSSQL then the string is formatted in the special unambigious MSSQL format
+# rather than ISO.
+#
+# Fortunately, there is a fairly simple workaround: we can just create a custom type
+# which "decorates" the built-in date type but does not actually modify its behaviour in
+# any way. SQLAlchemy will always defer to the custom type when working out how to
+# coerce expressions. Although we are in a sense "back" to having to use custom types,
+# we can use the same custom type across all databases; each dialect can still provide
+# its own custom implementation for built-in types, this trick just forces SQLAlchemy to
+# actually use them in all circumstances.
+
+
+class Date(sqlalchemy.types.TypeDecorator):
+    impl = sqlalchemy.types.Date
+    cache_ok = True
+
+
+class DateTime(sqlalchemy.types.TypeDecorator):
+    impl = sqlalchemy.types.DateTime
+    cache_ok = True
 
 
 TYPES_BY_NAME = {
-    "boolean": sqlalchemy.types.Boolean,
-    "date": sqlalchemy.types.Date,
-    "datetime": sqlalchemy.types.DateTime,
-    "float": sqlalchemy.types.Float,
-    "integer": sqlalchemy.types.Integer,
-    "varchar": sqlalchemy.types.Text,
-    "code": sqlalchemy.types.Text,
+    "boolean": Boolean,
```

Comment:
I wonder about leaving these "nude" SQLAlchemy types qualified by the module name rather than importing them into this module, so that the casual reader can see that there's something special about the ones that we've wrapped?

Reply:
I don't have a particularly firm view here. In my head the idea was that everywhere else in the codebase that touched SQLAlchemy types should import from here and it should be an alarm bell (although not necessarily an error) if we ever see them imported directly. I feel like the casual reader shouldn't have to care too much about the detail here beyond "always import types from this place". But open to arguments otherwise.

Reply:
I think you've accidentally explained why I'm wrong.

I was thinking of the casual reader of _this_ module. I had overlooked the fact that all the stuff from here is being imported elsewhere (and I think you're absolutely right that it's better to alias all the types via this module for that purpose). I suppose that in principle you _could_ use the qualified names here and yet still export the names from this module, but it would be weird. The reader elsewhere in the codebase is much more important.